source/iotagent-architecture.rst

```diff
@@ -0,0 +1,543 @@
+   compromised. For instance, if the private key associated to a particular
+   device is leaked, then all its messages should be ignored as there is no
+   guarantee that they came from that device.
+
+Each of these aspects will be detailed in the following sections.
+
+Device identity
+---------------
+
+The device identity verification is the starting point when dealing with
+communication security. This validation will indicate to the IoT agent if the
+device that opened the connection is whoever it says it is. Furthermore, the
+IoT agent must, once this validation succeeds, check whether this device can
+connect to it by checking its ID. This section will show how to do that.
+
+For connection-oriented protocols, the IoT agent should only accept connections
+for devices that have a certificate that was signed by an authority that is
+trusted by dojot. Once this certificate is valid, device identity can be
+checked in two forms:
+
+- Device ID encoded in certificate: although this is a less-reliable mechanism,
+  it allows greater flexibility using many devices in a controlled deployment.
+  This is based on setting the common name (CN certificate field) as dojot
+  device ID. Therefore, IoT agent should check whether this device exists or
+  not and allow or deny the connection right away depending on this
+  verification. The weak points of this mechanisms is that the device
+  certificate must be signed by dojot's internal CA (once there is a procedure
+  to sign only one certificate per device) and, if this certificate is valid,
+  then its ID must also be valid. If any other CA is used, then this mechanism
+  has no valid use.
+
+- IoT agent has all valid certificates: if an administrator wants to use an
+  external CA to sign all device certificates, then there is no actual control
+  of which device ID was used to generate a particular certificate. Therefore,
+  IoT agent must have all valid certificates properly mapped onto a device list
+  - this will guarantee that only one certificate is allowed to a particular
+  device and vice-versa.
+
+Using the first mechanism, the device (or an operator configuring a device for
+the first time) must call dojot CA to generate a signed certificate for itself.
+There is no further action for IoT agent to take as long as dojot CA is used.
+
+The second mechanism, however, requires that an IoT agent offer methods to
+manage certificates. The developer must take into account also that this IoT
+agent must be able to scale - these certificates must be accessible to all IoT
+agent instances, if allowed by deployment.
+
+Communication security
+----------------------
+
+With a valid certificate, a device can create a communication channel with
+dojot. For connection-oriented channels, this certificate should be used
+alongside cryptographic keys in order to provide an encrypted channel. For
+other channel types (such exchanging messages through a device gateway, such as
```

Comment:
( such as channels for exchaning ...  .... .... ....   .... , for example LoRa or sigfox